healthchecks: add DB, DB_NAME and support for several _FILE options

saserr opened the pull request:
## Description of changes

Added few additional options to services.healthchecks.settings:
- `DB`: with `sqlite` as default
- `DB_NAME`: with correct defaults based on `DB` value
- `_FILE` variants for several values that hold sensitive information in [Healthchecks configuration](https://healthchecks.io/docs/self_hosted_configuration/).

See the commit messages for more details.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

saserr committed:
nixos/healthchecks: define default DB_NAME for postgres and mysql

Previously, if someone changed DB to postgres or mysql and forgot to
change DB_NAME, services.healthchecks would have used the hardcoded path
that was meant for the sqlite as DB_NAME.

This change introduces DB and DB_NAME options in
services.healthchecks.settings.

saserr committed:
nixos/healthchecks: add EMAIL_HOST_PASSWORD_FILE option

This allows keeping EMAIL_HOST_PASSWORD out of /nix/store.

phaer reviewed (approved):
Thanks! 

Added a comment inline, but I otherwise this seems sensible to me. Not sure whether we have some kind of policy about which options we declare explicitly vs using cfg.settings, but as the latter still works i see no harm.

saserr commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -9,7 +9,7 @@ let
   boolToPython = b: if b then "True" else "False";
 
-  environment = { PYTHONPATH = pkg.pythonPath; } // cfg.settings;
+  environment = { PYTHONPATH = pkg.pythonPath; } // (filterAttrs (_: value: value != null) cfg.settings);
```
To be honest, I wasn't sure how to best solve this. Let me explain the issue and maybe you have a better suggestion.

The issue is that if there is no default value, then `cfg.settings` doesn't build because `EMAIL_HOST_PASSWORD_FILE` becomes required for user to set. But if there is a default value, then it needs to be a proper file path since it is being read in `local_setting.py`. So I thought of two possible solutions:
1. create an empty file in /nix/store as default so that an empty `EMAIL_HOST_PASSWORD` is set in local_setting.py. The issue with this is that it would always override `EMAIL_HOST_PASSWORD` if it is set in some other way. For example, if someone adds an additional environment file with `EMAIL_HOST_PASSWORD` in `systemd.services.healthchecks`).
2. make it `null` so that we can detect this value not being set and exclude it from the environment file. This one seemed like it has no issue.
3. correctly handle if `EMAIL_HOST_PASSWORD_FILE` is an empty string in `local_settings.py` and use an empty string as default value for the `EMAIL_HOST_PASSWORD_FILE` option. This also has no issue that I can think of and could be used as an alternative solution.

saserr replied:
Actually, I just checked and pyhton considers empty strings as `False` so 3rd options looks like a better option since it wouldn't require this filtering anymore. I'll update the CL.

saserr commented:
Thanks @phaer for such a speedy review and a great comment that forced me to simplify the code.

@SuperSandro2000 is this something you would be able to merge? Please feel free to review the code and comment as well since this is my first contribution 😃

SuperSandro2000 commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -143,8 +151,34 @@ in
+            default =
+              if settings.config.DB == "sqlite"
+              then "${cfg.dataDir}/healthchecks.sqlite"
+              else "hc";
```
Was that the default before? I think using healthcheck would be more descriptive.

phaer replied:
old default is explained in the commit message https://github.com/NixOS/nixpkgs/pull/251323/commits/ef9725f01fcc7cd49bb14e13e5ee43e1f687ab5b

saserr replied:
No, there was no default because sqlite path was used as the default value even if DB was postgres / mysql. I took this default from Healthchecks documentation (see https://healthchecks.io/docs/self_hosted_configuration/#DB_NAME).

What is the convention in nixpkgs: to use a better default value or to use the package's defaults?

SuperSandro2000 replied:
> What is the convention in nixpkgs: to use a better default value or to use the package's defaults?

usually we stick to upstream unless there are good reasons not to. Both are probably fine here.

phaer commented on pkgs/servers/web-apps/healthchecks/default.nix:
```diff
@@ -41,11 +41,18 @@ py.pkgs.buildPythonApplication rec {
+
     STATIC_ROOT = os.getenv("STATIC_ROOT")
+
     SECRET_KEY_FILE = os.getenv("SECRET_KEY_FILE")
     if SECRET_KEY_FILE:
         with open(SECRET_KEY_FILE, "r") as file:
             SECRET_KEY = file.readline()
+
+    EMAIL_HOST_PASSWORD_FILE = os.getenv("EMAIL_HOST_PASSWORD_FILE")
+    if EMAIL_HOST_PASSWORD_FILE:
+        with open(EMAIL_HOST_PASSWORD_FILE, "r") as file:
+            EMAIL_HOST_PASSWORD = file.readline()
```
Don't see what could be gained from deduplication with n=2, 4 lines each. This code is clear and easy to cut, copy, paste if needed.

saserr replied:
I am happy to make this a for-loop.

There is actually quite a few secrets in Healthchecks configuration that should probably be configurable as files (see https://healthchecks.io/docs/self_hosted_configuration/). I personally don't need them, but I can add them to `healthchecks.settings` if you believe it is worth it.

saserr replied:
I added support for all variables that hold sensitive information. Please let me know if this looks ok.

phaer commented on nixos/modules/services/web-apps/healthchecks.nix:
```diff
@@ -118,6 +117,15 @@ in
             apply = lib.concatStringsSep ",";
           };
 
+          STATIC_ROOT = mkOption {
+            type = types.str;
+            default = cfg.dataDir + "/static";
+            defaultText = lib.literalExpression ''
+              config.${opt.dataDir} + "/static"
+            '';
+            description = lib.mdDoc "State directory for dynamically generated static files.";
+          };
```
> Wouldn't it be easier to override the package and change the files in there?

The files are generated dynamically by the manage.py script, so I fail to see how overriding would improve things here?

saserr replied:
This was already overridable through `healthchecks.settings` because it is being read from environment in local_setting.py and the default path was defined in the environment above. I just moved it here and documented it. Note that it was already mentioned in description for `healthchecks.settings` too.

I am happy to remove this change if you think it creates unnecessary confusion. I actually don't need it and only wanted to clean things up a bit.

saserr replied:
Sounds like there is a consensus that this change is not needed. Dropped this commit.

saserr committed:
nixos/healthchecks: enable _FILE variants for all secrets

This change enables _FILE variants for all secrets in Healthchecks
configuration so they can be read from a file and not stored in
/nix/store.

In particular, it adds support for these secrets:
DB_PASSWORD, DISCORD_CLIENT_SECRET, EMAIL_HOST_PASSWORD,
LINENOTIFY_CLIENT_SECRET, MATRIX_ACCESS_TOKEN, PD_APP_ID,
PUSHBULLET_CLIENT_SECRET, PUSHOVER_API_TOKEN, S3_SECRET_KEY, SECRET_KEY,
SLACK_CLIENT_SECRET, TELEGRAM_TOKEN, TRELLO_APP_KEY, and TWILIO_AUTH.

saserr commented:
@phaer do you have time to re-review this change after 5595c7aee4a0bea5eee544c5dc5f1a8fe27a2ace?

phaer commented:
@saserr I can do that, but I can't merge (not a committer)

saserr commented:
> @saserr I can do that, but I can't merge (not a committer)

@phaer no worries and I really appreciate all your help with this PR.

@SuperSandro2000 sorry to bother you again, but are you able to merge this PR?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1079

phaer commented:
> > @saserr I can do that, but I can't merge (not a committer)
> @phaer no worries and I really appreciate all your help with this PR.

Thank *you* for contributing! :) There's a dedicated discourse thread for already-reviewed PRs. I mentioned this one, so there's a link above my post.

Lassulus commented:
Result of `nixpkgs-review pr 251323` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>healthchecks</li>
  </ul>
</details>
